avoid inconsistent state in Block

gulige opened the pull request:
In the member function insert(position, elem) to class Block, if function checkColumnStructure<void> throws an exception for columns with same name but unidentical structure, it will result in the mismatch between index_by_name and data (indexes changed while data not). However we could not presume upon that caller will not use this block object anymore. So it seems better to guarantee consistency innerly.

Changelog category:
- Not for changelog

gulige committed:
avoid inconsistent state in Block

In the member function insert(position, elem) to class Block, if function checkColumnStructure<void> throws an exception for columns with same name but unidentical structure, it will result in the mismatch between index_by_name and data. However we could not presume upon that caller will not use this block object anymore. So it seems better to guarantee consistency innerly.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=32641) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=32641) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=32641) it.</sub>

gulige committed:
Merge branch 'ClickHouse:master' into master

alesapin commented:
@mergifyio update

mergify[bot] committed:
Merge branch 'master' into master

alexey-milovidov commented on src/Core/Block.cpp:
```diff
@@ -139,15 +139,17 @@ void Block::insert(size_t position, ColumnWithTypeAndName elem)
-            ++name_pos.second;
-
-    auto [it, inserted] = index_by_name.emplace(elem.name, position);
+    auto [it, inserted] = index_by_name.emplace(elem.name, (size_t)-1);
```
This will give compiler warning about "C-style cast".

gulige replied:
That's correct! I will use std::npos to replace (size_t)-1. Thanks for review.

zuochunwei replied:
yes，use std::npos is better!

gulige committed:
to clear up compiler warning about "C-style cast".

use std::npos to replace (size_t)-1

kitaisreal commented on src/Core/Block.cpp:
```diff
@@ -139,15 +139,17 @@ void Block::insert(size_t position, ColumnWithTypeAndName elem)
-            ++name_pos.second;
-
-    auto [it, inserted] = index_by_name.emplace(elem.name, position);
+    auto [it, inserted] = index_by_name.emplace(elem.name, std::npos);
```
Better to not use emplace with std::npos, just find element, if it already exists then checkColumnStructure, after that we can safely update index_by_name and insert element.

gulige replied:
Yes, better way. I will refactor it as you said.

gulige replied:
The current code changes are supposed to comply with the original code semantically and can solve the potential inconsistent state in Block. However, I need to point out, for the case of  columns with both identical name and identical structure, the new index_by_name entry still has NOT been established for conflicting name. This is tolerable because both old index_by_name entry and old data entry still exist. It has no other side effects except one new data entry being inserted with no index pointing to it. It is somewhat complicated to completely get rid of old index_by_name entry and old data and then replace them with the corresponding new ones. And I also don't know if this logic is the right one you want.

gulige committed:
code refactoring as @kitaisreal suggested

Better to not use emplace with std::npos, just find element, if it already exists then checkColumnStructure, after that we can safely update index_by_name and insert element.

gulige committed:
missing code refactoring as @kitaisreal suggested

gulige committed:
fix code refactoring as @kitaisreal suggested

gulige committed:
Merge branch 'ClickHouse:master' into master

alexey-milovidov reviewed (changes requested):
We can rewrite this code to have only one lookup in the map:
- emplace before modifying positions;
- when modifying positions, skip the inserted element by comparing iterators.

gulige committed:
Merge branch 'ClickHouse:master' into master

gulige committed:
code refactoring as @alexey-milovidov suggested

alexey-milovidov commented on src/Core/Block.cpp:
```diff
@@ -145,15 +145,17 @@ void Block::insert(size_t position, ColumnWithTypeAndName elem)
 
+    for (auto itr = index_by_name.begin(); itr != index_by_name.end(); ++itr)
+    {
+        if (itr->second >= position && (!inserted || itr != it))
```
But we already checked for `!inserted` above.

gulige replied:
Because this is the common logic for the cases:
1) inserted success
2) not inserted, but checkColumnStructure() was ok and didn't throw exception

alexey-milovidov replied:
Ok.

alexey-milovidov committed:
Update Block.cpp

alexey-milovidov commented:
Ok. Please keep in mind that LOGICAL_ERROR exception is indicating error in your code. It should not happen in production. And in the case of this error it does not really matter that the state will remain consistent.
